weed/storage/needle_map_leveldb.go

```diff
@@ -235,3 +235,100 @@ func (m *LevelDbNeedleMap) Destroy() error {
 	os.Remove(m.indexFile.Name())
 	return os.RemoveAll(m.dbFileName)
 }
+
+func (m *LevelDbNeedleMap) UpdateNeedleMap(v *Volume, indexFile *os.File, opts *opt.Options) error {
+	if v.nm != nil {
+		v.nm.Close()
+		v.nm = nil
+	}
+	defer func() {
+		if v.tmpNm != nil {
+			v.tmpNm.Close()
+			v.tmpNm = nil
+		}
+	}()
+	levelDbFile := v.FileName(".ldb")
+	m.indexFile = indexFile
+	err := os.RemoveAll(levelDbFile)
+	if err != nil {
+		return err
+	}
+	if err = os.Rename(v.FileName(".cpldb"), levelDbFile); err != nil {
+		return fmt.Errorf("rename %s: %v", levelDbFile, err)
+	}
+
+	db, err := leveldb.OpenFile(levelDbFile, opts)
+	if err != nil {
+		if errors.IsCorrupted(err) {
+			db, err = leveldb.RecoverFile(levelDbFile, opts)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	stat, e := indexFile.Stat()
+	if e != nil {
+		glog.Fatalf("stat file %s: %v", indexFile.Name(), e)
+		indexFile.Close()
+		db.Close()
+		return e
+	}
+	m.indexFileOffset = stat.Size()
+	m.recordCount = uint64(stat.Size() / types.NeedleMapEntrySize)
+	m.db = db
+
+	v.nm = m
+	v.tmpNm = nil
+	return e
+}
+
+func (m *LevelDbNeedleMap) DoOffsetLoading(v *Volume, indexFile *os.File, startFrom uint64) error {
+	glog.V(0).Infof("loading idx to leveldb from offset %d for file: %s", startFrom, indexFile.Name())
+	dbFileName := v.FileName(".cpldb")
+	db, err := leveldb.OpenFile(dbFileName, nil)
+	if err != nil {
+		if errors.IsCorrupted(err) {
+			db, err = leveldb.RecoverFile(dbFileName, nil)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	defer db.Close()
+
+	//set watermark
+	stat, e := indexFile.Stat()
+	if e != nil {
+		return e
+	}
+	watermark := stat.Size() / watermarkBatchSize
+	e = setWatermark(db, uint64(watermark))
```

Comment:
Should it set the watermark after successfully load completion time, not before it starts?

Reply:
I think the order is not important.
No matter set it on the begaining or the end, it equals to entry counts / 10000. And entry counts stays same when begain to load volume.

Reply:
I may make a mistake. Wait a minute, I'll fix it

Reply:
what if the loading has error?

Reply:
Remove the .cpldb folder if error happens

Reply:
Done